### PR TITLE
[240] Importation utilisateur fichier préfecture

### PR DESCRIPTION
**Pull request by jdauphant**

### CODE REVIEW

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Coverage increased per file
===========================
+ app/csv/GroupImport.scala  4
+ app/csv/package.scala  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 25
- Added 27
           

Coverage increased per file
===========================
+ app/csv/GroupImport.scala  4
+ app/csv/package.scala  21
         

Coverage decreased per file
===========================
- app/csv/UserImport.scala  -19
         

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 30
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -1
         

Clones added
============
- test/csv/CSVSpec.scala  4
- app/controllers/UserController.scala  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 30
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -1
         

Clones added
============
- test/csv/CSVSpec.scala  4
- app/controllers/UserController.scala  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 30
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -1
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 30
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -1
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 30
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -1
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 26
- Added 31
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Complexity decreasing per file
==============================
+ app/csv/package.scala  -1
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  4
- app/controllers/UserController.scala  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by lpereir4** [comment]
- Le bouton d'import depuis un fichier csv se trouve en haut à gauche.

Screenshot 2019-12-05 at 09 44 32
- Choisissez le séparateur à utiliser et collez le csv dans la zone de texte.
Screenshot 2019-12-05 at 09 45 27
- Vérifiez le contenu importé. Le choix du département est global. Il vous est possible d'assigner toutes les personnes à la fonction d'instructeur.
Screenshot 2019-12-05 at 09 47 01

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Coverage increased per file
===========================
+ app/csv/package.scala  29
         

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  4
- app/controllers/UserController.scala  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- app/controllers/UserController.scala  2
- test/csv/CSVSpec.scala  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Coverage increased per file
===========================
+ app/csv/package.scala  29
         

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  4
- app/controllers/UserController.scala  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  4
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  4
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 42
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  4
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 46
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  7
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 46
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  7
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 46
           

Coverage increased per file
===========================
+ app/csv/package.scala  29
         

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  7
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 48
           

Coverage increased per file
===========================
+ app/csv/package.scala  29
         

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  7
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 27
- Added 48
           

Complexity increasing per file
==============================
- public/javascripts/csv-import.js  2
         

Clones added
============
- test/csv/CSVSpec.scala  7
         

Clones removed
==============
+ app/controllers/UserController.scala  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/aplus/aplus/pullRequest?prid=4608500&bid=15462054)

File: app/controllers/UserController.scala [hunk 1]
```diff
@@ -173,19 +174,23 @@ case class UserController @Inject()(loginAction: LoginAction,
     asAdmin { () =>
       "IMPORT_GROUP_UNAUTHORIZED" -> "Accès non autorisé pour importer les utilisateurs"
     } { () =>
-      csv.sectionsForm.bindFromRequest.fold({ missFilledForm =>
+      csv.sectionsForm(request.currentUser.id).bindFromRequest.fold({ missFilledForm =>
         val cleanedForm = missFilledForm.copy(data = missFilledForm.data.filter({ case (_, v) => v.nonEmpty }))
-        BadRequest(views.html.reviewUsersImport(request.currentUser, request.currentArea)(cleanedForm))
-      }, { sections =>
+        BadRequest(views.html.reviewUsersImport(request.currentUser, request.currentArea)(cleanedForm, Nil))
+      }, { case (sections, territoryId) =>
```
Comment (jdauphant) [inline app/controllers/UserController.scala]: Dans le code c'est Area et non Territory qui est utilisé.
(est-ce qu'on peut avoir comme convention de bien garder les mêmes terme ? ou de se mettre d'accord pour les changer ?)

Comment (lpereir4) [reply app/controllers/UserController.scala]: Choisissons area.

File: app/controllers/UserController.scala [hunk 2]
```diff
@@ -331,7 +336,8 @@ case class UserController @Inject()(loginAction: LoginAction,
         "groupIds" -> default(list(uuid), List()),
         "delegations" -> ignored(Map[String, String]()),
         "cguAcceptationDate" -> optional(ignored(Time.now())),
-        "newsletterAcceptationDate" -> optional(ignored(Time.now()))
+        "newsletterAcceptationDate" -> optional(ignored(Time.now())),
```
Comment (jdauphant) [inline app/controllers/UserController.scala]: Il ne faut pas mettre de valeur, ça déclenche le panneau de CGU et de validation de newsletter.
ignored(None)

Comment (jdauphant) [reply app/controllers/UserController.scala]: Ah ça ne doit pas le sauvergarder en fait. J'ai du faire ce code. Peut être rajouter un commentaire pour informer du comportement.

File: app/csv/package.scala [hunk 3]
```diff
@@ -1,138 +1,206 @@
...
+      .transform[Boolean](os => os.exists(s => Operators.not(s.isEmpty)), helper => if (helper) Some(GROUP_MANAGER.prefixes(0)) else None),
+
+    "disabled" -> ignored(false),
+    "expert" -> ignored(false),
+    "groupIds" -> default(list(uuid), List()),
+    "delegations" -> default(seq(tuple("name" -> nonEmptyText, "email" -> email))
+      .transform[Map[String, String]](_.toMap, _.toSeq), Map.empty[String, String]),
+
+    "cguAcceptationDate" -> optional(ignored(Time.now())),
+    "newsletterAcceptationDate" -> optional(ignored(Time.now()))
```
Comment (jdauphant) [inline app/csv/package.scala]: Il faut importer la cguAcceptationDate et la newsletterAcceptationDate à None pour que ça déclenche la page de validation CGU.

File: app/views/importUsers.scala.html [hunk 4]
```diff
@@ -9,10 +9,15 @@
...
+                <em>Séparateur</em>
+                <input type="radio" name="separator" value=";" id="semicolon" checked><label for="semicolon">Point-virgule</label>
+                <input type="radio" name="separator" value="," id="comma"><label for="comma">Virgule</label>
+            </div>
+            <p>@csv.USER_HEADER@csv.SEPARATOR@csv.GROUP_HEADER</p>
             @if(errors.nonEmpty) {
-                <div style="color: red; font-weight: bold;">@errors.mkString(", ")</div>
+                <div style="color: red; font-weight: bold;">@errors.map(_.message)</div>
             }
             <textarea label="Contenu du CSV"
```
Comment (jdauphant) [inline app/views/importUsers.scala.html]: On a peut être moyen de mettre la textarea plus haute. (tous le monde n'a pas chrome qui permet ça facilement.

File: app/views/reviewUsersImport.scala.html [hunk 5]
```diff
@@ -1,6 +1,6 @@
 @import models._
 @import extentions.MDLForms._
-@(currentUser: User, area: Area)(form: Form[List[csv.Section]])(implicit webJarsUtil: org.webjars.play.WebJarsUtil, flash: Flash, messagesProvider: MessagesProvider, request: RequestHeader)
+@(currentUser: User, area: Area)(form: Form[(List[csv.Section], java.util.UUID)], errors: List[String])(implicit webJarsUtil: org.webjars.play.WebJarsUtil, flash: Flash, messagesProvider: MessagesProvider, request: RequestHeader)
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: Pour les errors, je les inject dans le form (j'ai fait ça pour l'autre vue d'ajout utilisateur). Ça peut simplifier ton api.

File: app/views/reviewUsersImport.scala.html [hunk 6]
```diff
@@ -16,48 +16,42 @@
...
                  type="checkbox"
-                 name="@user(csv.GROUP_MANAGER_HEADER_PREFIX).name"
-                 id="@user(csv.GROUP_MANAGER_HEADER_PREFIX).id"
-                 value="@user(csv.GROUP_MANAGER_HEADER_PREFIX).value.getOrElse("")"
-                 @{if(!user(csv.GROUP_MANAGER_HEADER_PREFIX).value.getOrElse("").isEmpty) "checked"} />
-        }
+                 name="@user(csv.INSTRUCTOR.key).name"
+                 id="@user(csv.INSTRUCTOR.key).id"
+                 value="@user(csv.INSTRUCTOR.key).value.getOrElse("")"
+                 @{if(!user(csv.INSTRUCTOR.key).value.getOrElse("").isEmpty) "checked"} />
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: Tu as vu l'helper de checkbox ?
@helper.checkbox(userField("helper"), 'type -> "checkbox", 'label -> "Aidant", 'class -> "mdl-checkbox__input") 

Il s'occupe de générer tous ça.

File: app/views/reviewUsersImport.scala.html [hunk 7]
```diff
@@ -66,39 +60,39 @@
...
 
 @main(currentUser, area, false)("Revue de votre import CSV") {
+  <script src='@routes.Assets.versioned("javascripts/csv-import.js")'></script>
 } {
   @helper.form(routes.UserController.importUsers(), 'method -> "post", 'class -> "mdl-cell mdl-cell--12-col") {
   @helper.CSRF.formField
-    @if(form.hasGlobalErrors) {
-      <div style="color: red; font-weight: bold;">@form.globalErrors.mkString(", ")</div>
+    @for(error <- errors) {
+      <div style="color: red; font-weight: bold;">@error</div>
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: Pas rapport à la remarque plus haut, c'est ici qu'on aurait pu réinjecter les erreurs dans le form.

**Pull request merged**